### ISSUE TITLE
Reconfiguring tarantool with replication_connect_quorum=0 hangs

### ISSUE THREAD

**Issue by rosik** [tarantool/tarantool#3760]
Tarantool version: `Tarantool 1.10.2-0-gc0d8063b6`

OS version: `Ubuntu bionic 18.04, 4.15.0-36-generic`

Steps to reproduce:

1. Bootstrap the first instance:
```console
tarantool> box.cfg({listen=3301})
tarantool> box.schema.user.grant('guest', 'replication')
```

2. Bootstrap the second instance
```console
tarantool> box.cfg({listen=3302, replication={'localhost:3301'}})
```

3. Shutdown the first instance

4. Reconfigure second instance again with `replication_connect_timeout` and `replication_connect_quorum` specified
```console
tarantool> box.cfg({replication={'localhost:3301'}, replication_connect_quorum=0, replication_connect_timeout=4})
2018-10-23 20:10:46.697 [17197] main/101/interactive I> set 'replication_connect_quorum' configuration option to 0
2018-10-23 20:10:46.697 [17197] main/101/interactive I> set 'replication_connect_timeout' configuration option to 4
2018-10-23 20:10:46.698 [17197] main/101/interactive I> connecting to 1 replicas
2018-10-23 20:10:46.698 [17197] main/114/applier/localhost:3301 I> can't connect to master
2018-10-23 20:10:46.698 [17197] main/114/applier/localhost:3301 coio.cc:106 !> SystemError connect, called on fd 12, aka 127.0.0.1:45096: Connection refused
2018-10-23 20:10:46.698 [17197] main/114/applier/localhost:3301 I> will retry every 1.00 second
2018-10-23 20:10:50.696 [17197] main/101/interactive C> failed to connect to 1 out of 1 replicas
2018-10-23 20:10:50.696 [17197] main/101/interactive I> set 'replication' configuration option to ["localhost:3301"]
---
...
```

**Result:** tarantool hangs for `replication_connect_timeout` seconds

**Comment by kyukhin** [comment]
What do expect?

**Comment by rosik** [comment]
The [doc](https://tarantool.io/en/doc/1.9/book/replication/repl_add_instances/#orphan-status) says:

```
Situation 3: configuration update
...
1. Try to connect to all nodes from box.cfg.replication,
or to the number of nodes required by replication_connect_quorum,
within the time period specified in replication_connect_timeout.
```

I like this wording, and expect that box.cfg would return as soon as it reaches the quorum. In my case, quorum=0, it means it should return instantly.

**Comment by Gerold103** [comment]
The provided test now does not hang. Perhaps because now we don't try to reconnect to replicasets, if box.cfg.replication is not changed. But this hangs:
First instance:
```Lua
box.cfg({listen=3301})
box.schema.user.grant('guest', 'replication')
```
Second instance:
```Lua
box.cfg({listen=3302, replication={'localhost:3301'}, replication_connect_quorum=0, replication_connect_timeout=4})
```

Now shutdown the first instance, and reconfigure the second instance:
```Lua
box.cfg({replication={}})
box.cfg({replication={'localhost:3301'}})
```

**Comment by Gerold103** [comment]
The problem is deeper.
Quorum is not respected regardless of its value.
Instance 1:
```Lua
box.cfg({listen=3301, replication={'localhost:3301', 'localhost:3302', 'localhost:3303'}})
box.schema.user.grant('guest', 'replication')
```
Instance 2:
```Lua
box.cfg({listen=3302, replication={'localhost:3301', 'localhost:3302', 'localhost:3303'}})
box.schema.user.grant('guest', 'replication')
```
Instance 3:
```Lua
box.cfg({listen=3303, replication={'localhost:3301', 'localhost:3302', 'localhost:3303'}, replication_connect_quorum=1, replication_connect_timeout=4})
```

Shutdown instance 1 and reconfigure instance 3:
```Lua
box.cfg({replication={}})
box.cfg({replication={'localhost:3301', 'localhost:3302', 'localhost:3303'}})
```
This hangs for replication connect timeout even though quorum 1 is satisfied earlier.

**Comment by rosik** [comment]
Please, cherry-pick it to 1.10

**Comment by rosik** [comment]
Thanks

**Comment by Totktonada** [comment]
(Just to clarify it for myself.) AFAIU c6bea65f8ef5f6c737cf70c0127189d0ebcbc36e fixes a most part of the problem, then 00c6c4377514abfd6e1cce11c3b4d7406b29db96 (#4433) ends it.
